cc-snapshot

```diff
@@ -308,51 +291,70 @@ if [ $DISTRO = $UBUNTU ]; then
+    LABEL="$(ls -1 /dev/disk/by-label)"
+    if [[ "$(wc -l <<<"$LABEL")" != "1" ]]; then
+      >&2 echo "Unable to auto-detect disk label: found >1 labeled disks:"
+      >&2 echo "$(sed 's/\n/ - /g' <<<"$LABEL")"
+      >&2 echo "Please try again while explicitly setting the LABEL"
+      >&2 echo "env variable to the desired disk label."
+      exit 1
+    fi
+  fi
+}
 
-# This will take 3 to 5 minutes. Next, convert the tar file into a qcow2 image
-# (if you don't want to use the XFS file system, you can replace xfs by ext4):
-$VIRT_MAKE_FS --format="$CC_SNAPSHOT_DISK_FORMAT" \
-  --type="$FS" \
-  --label="$LABEL" \
-  "$CC_SNAPSHOT_TAR_PATH" \
-  "$CC_SNAPSHOT_CONVERTED_PATH"
-
-# Our ARM64 images use U-Boot, so don't install GRUB
-if [ "$PARTITION_IMAGE" == false ] && [ "$ARCH" != "aarch64" ]; then
-  if [ $DISTRO = $CENTOS ]; then
+if [ $DISTRO = $CENTOS ]; then
+    # TODO: UEFI support for centos
+
+    # This will take 3 to 5 minutes. Next, convert the tar file into a qcow2 image
+    # (if you don't want to use the XFS file system, you can replace xfs by ext4):
+    virt-make-fs --partition --format="$CC_SNAPSHOT_DISK_FORMAT" \
+    --type="$FS" \
+    --label="$LABEL" \
+    "$CC_SNAPSHOT_TAR_PATH" \
+    "$CC_SNAPSHOT_CONVERTED_PATH"
     # and looking at the name of the file in that directory. Next ensure that the GRUB bootloader is present in the image:
     virt-customize -a $CC_SNAPSHOT_CONVERTED_PATH --run-command 'grub2-install /dev/sda && grub2-mkconfig -o /boot/grub2/grub.cfg'
-  fi
+fi
 
-  if [ $DISTRO = $UBUNTU ]; then
+if [ $DISTRO = $UBUNTU ]; then
     # Next ensure that the GRUB bootloader is present in the image:
-    guestfish -a $CC_SNAPSHOT_CONVERTED_PATH -i sh 'grub-install /dev/sda && grub-mkconfig -o /boot/grub/grub.cfg'
-  fi
+    ARCH=$(uname -i)
+    case "$ARCH" in
+        "x86_64") GRUB_INSTALL_TARGET="x86_64-efi"
+        ;;
+        "aarch64"|"arm64") GRUB_INSTALL_TARGET="arm64-efi"
+        ;;
+        *)
+            echo "$ARCH unsupported!"
+            exit 1
+        ;;
+    esac
+    CC_SNAPSHOT_CONVERTED_PATH=/tmp/snapshot.img
+    filesystem_size=$(du -d 0 -m --exclude-from=$EXCLUDE_FROM / |  awk -F" "  '{ print $1 }')
+    total_disk_size=$(($filesystem_size + 550 + 8))
+
+    guestfish -N $CC_SNAPSHOT_CONVERTED_PATH=disk:"$total_disk_size"M -- \
+    part-init /dev/sda gpt : \
+    part-add /dev/sda primary 2048 1128447 : \ # 550M for efi partition
+    part-add /dev/sda primary 1128448 1144831 : \ # 8M for bios partition
+    part-add /dev/sda primary 1144832 -40 : \
+    mkfs $FS /dev/sda3 label:$LABEL : \
+    mkfs vfat /dev/sda1 : \
+    mount /dev/sda3 / : tar-in $CC_SNAPSHOT_TAR_PATH / : mount /dev/sda1 /boot/efi : \
+    sh "parted /dev/sda set 2 bios_grub on" : \
+    sh "parted /dev/sda set 1 esp on" : \
+    sh "grub-install --target=$GRUB_INSTALL_TARGET --efi-directory=/boot/efi --removable && grub-install /dev/sda && grub-mkconfig -o /boot/grub/grub.cfg"
```

Comment:
This looks like the "magic" that makes the grub install work now?

Reply:
yes :)